File: napkin/patch/patchscene.py
import math

from PyQt5.QtCore import pyqtSignal, Qt, QPointF
from PyQt5.QtGui import QPen
from PyQt5.QtWidgets import QGraphicsScene, QApplication

from patch.layeritem import LayerItem
from patch.inputoutputnodeitem import InputOutputNodeItem, NodeItem
from patch.patchutils import _getObjectEditorPos
from patch._plugitem import _PlugItem
from patch.edgeitem import EdgeItem
from patch.wirepreview import WirePreview

_typeFilter = lambda m, t: isinstance(m, t)


class PatchScene(QGraphicsScene):
    nodeSelectionChanged = pyqtSignal(list)

    def __init__(self):
        """
        @type ctx: AppContext
        """
        super(PatchScene, self).__init__()
        self.__wireIsOutput = False

        self.__nodeLayer = LayerItem()
        self.addItem(self.__nodeLayer)
        self.__edgeLayer = LayerItem()
        self.addItem(self.__edgeLayer)
        self.__interactionLayer = LayerItem()
        self.__interactionLayer.setAcceptedMouseButtons(Qt.NoButton)
        self.__interactionLayer.setAcceptHoverEvents(False)
        self.__interactionLayer.setAcceptTouchEvents(False)
        self.__interactionLayer.setEnabled(False)
        self.addItem(self.__interactionLayer)

        self.__previewWire = WirePreview(self.__interactionLayer)
        self.__previewWire.setEnabled(False)
        self.__previewWire.setParentItem(self.__interactionLayer)
        self.__previewWire.setVisible(False)

        self.__isWiring = False

        # self.setSceneRect(-1000, -1000, 1000, 1000)
        self.selectionChanged.connect(self.__onSelectionChanged)

        for opItem in self.nodes():
            for inPlug in opItem.operator().inputPlugs():
                con = inPlug.connection()
                if con:
                    self.addEdge(con, inPlug)

        self.changed.connect(self.__sceneChanged)

    def __sceneChanged(self, regions):
        self.ensureLargeEnoughSceneRect()

    def ensureLargeEnoughSceneRect(self):
        """Enlarge scenerect so dragging nodes around doesn't feel so weird"""
        margin = 1000
        adjusted = self.itemsBoundingRect().adjusted(-margin, -margin, margin, margin)
        self.setSceneRect(self.sceneRect().united(adjusted))

    def addNode(self, node: NodeItem):
        self.addItem(node)

    def startDragConnection(self, pinItem):
        self.hideIncompaticlePlugs(pinItem.plugItem())
        self.__wireIsOutput = not pinItem.plugItem().isInput()
        if self.__wireIsOutput:
            self.__previewWire.srcPin = pinItem
            self.__previewWire.dstPos = pinItem.attachPos()
        else:
            self.__previewWire.dstPin = pinItem
            self.__previewWire.srcPos = pinItem.attachPos()
        self.__previewWire.setVisible(True)
        self.__isWiring = True

    def updateDragConnection(self, pt):
        if self.__wireIsOutput:
            self.__previewWire.dstPos = pt
        else:
            self.__previewWire.srcPos = pt
        self.__previewWire.updatePath()

    def stopDragConnection(self, pinItem):
        if not self.__isWiring:
            return
        if pinItem:
            if self.__wireIsOutput:
                srcPlugItem = self.__previewWire.srcPin.plugItem()
                dstPlugItem = pinItem.plugItem()
            else:
                srcPlugItem = pinItem.plugItem()
                dstPlugItem = self.__previewWire.dstPin.plugItem()

            if srcPlugItem != dstPlugItem:
                srcPlug = srcPlugItem.plug()
                dstPlug = dstPlugItem.plug()
                dstPlug.connectTo(srcPlug)

        self.__previewWire.srcPin = None
        self.__previewWire.dstPin = None
        self.__previewWire.setVisible(False)
        self.__isWiring = False
        self.showAllPlugs()

    def drawBackground(self, painter, rect):
        spacing = 50
        xmin = int(math.floor(rect.left() / spacing))
        xmax = int(math.ceil(rect.right() / spacing))
        ymin = int(math.floor(rect.top() / spacing))
        ymax = int(math.ceil(rect.bottom() / spacing))

        pen = QPen()
        pen.setWidth(0)
        pen.setColor(QApplication.palette().base().color().darker(110))
        painter.setPen(pen)

        for x in range(xmin, xmax):
            painter.drawLine(QPointF(x, ymin) * spacing,
                             QPointF(x, ymax) * spacing)
        for y in range(ymin, ymax):
            painter.drawLine(QPointF(xmin, y) * spacing,
                             QPointF(xmax, y) * spacing)

    def nodes(self):
        return filter(lambda m: isinstance(m, NodeItem),
                      self.__edgeLayer.childItems())

    def edges(self):
        return filter(lambda m: isinstance(m, EdgeItem),
                      self.__edgeLayer.childItems())

    def hideIncompaticlePlugs(self, src):
        for opItem in self.nodes():
            opItem.hideIncompatiblePlugs(src)

    def showAllPlugs(self):
        for opItem in self.nodes():
            opItem.showAllPlugs()

    def selectedNodes(self):
        return filter(lambda m: isinstance(m, NodeItem), self.selectedItems())

    def selectedEdges(self):
        return filter(lambda m: isinstance(m, EdgeItem), self.selectedItems())

    def findEdge(self, plugA: _PlugItem, plugB: _PlugItem):
        for item in self.edges():
            if item.srcPin.plugItem() == plugA and item.dstPin.plugItem() == plugB:
                return item
            if item.dstPin.plugItem() == plugB and item.dstPin.plugItem() == plugA:
                return item

    def dragConnectionSource(self):
        if self.__previewWire.srcPin:
            return self.__previewWire.srcPin.plugItem().plug()
        if self.__previewWire.dstPin:
            return self.__previewWire.dstPin.plugItem().plug()
        return None

    def __onSelectionChanged(self):
        self.nodeSelectionChanged.emit(list(self.selectedNodes()))

    def addEdge(self, srcPlug: _PlugItem, dstPlug: _PlugItem):
        srcPlugItem = self.findOperatorItem(srcPlug.parent()).findPlugItem(
            srcPlug)
        dstPlugItem = self.findOperatorItem(dstPlug.parent()).findPlugItem(
            dstPlug)

        wire = EdgeItem(srcPlugItem.pin(), dstPlugItem.pin())
        wire.setParentItem(self.__edgeLayer)

    def removeEdge(self, srcPlug, dstPlug):
        wire = self.findEdge(srcPlug, dstPlug)
        assert (wire)
        self.__removeItem(wire)

    def __onOperatorAdded(self, op):
        """
        @type op: nap.Operator
        """
        item = InputOutputNodeItem(self.__nodeLayer, op)
        item.moved.connect(self.__updateSceneRect)
        item.setParentItem(self.__nodeLayer)
        item.plugConnected.connect(self.addEdge)
        item.plugDisconnected.connect(self.removeEdge)
        item.setPos(_getObjectEditorPos(op))
        self.__updateSceneRect()

    def __onOperatorRemoved(self, op):
        self.__removeItem(self.findOperatorItem(op))

    def __onOperatorChanged(self, op):
        self.findOperatorItem(op).setPos(_getObjectEditorPos(op))

    def __removeItem(self, item):
        self.removeItem(item)
        del item

    def __updateSceneRect(self):
        self.setSceneRect(
            self.itemsBoundingRect().adjusted(-1000, -1000, 1000, 1000))

File: napkin/src_old/patchpanel/patchscene.cpp
#include "patchscene.h"
#include <nap/operator.h>

template <typename T>
inline QList<T> filter(const QList<QGraphicsItem*>& list)
{
	QList<T> filtered;
	for (QGraphicsItem* item : list) {
		T t = qgraphicsitem_cast<T>(item);
		if (t) filtered.append(t);
	}
	return filtered;
}


PatchScene::PatchScene(nap::Patch& patch) : QGraphicsScene(nullptr), mPatch(patch)
{
	mIsWiring = false;

	mOperatorLayer = new Layer();
	mWireLayer = new Layer();
	mInteractionLayer = new Layer();
	mInteractionLayer->setAcceptedMouseButtons(Qt::NoButton);
	mInteractionLayer->setAcceptHoverEvents(false);
	mInteractionLayer->setAcceptTouchEvents(false);
	mInteractionLayer->setEnabled(false);

	mPreviewWire = new WirePreview(mInteractionLayer);
	mPreviewWire->setEnabled(false);
	mPreviewWire->setParentItem(mInteractionLayer);
	mPreviewWire->setVisible(false);

	addItem(mWireLayer);
	addItem(mOperatorLayer);
	addItem(mInteractionLayer);

	setSceneRect(-1000, -1000, 2000, 2000);

	connect(this, SIGNAL(selectionChanged()), this, SLOT(onSelectionChanged()));

	mPatch.operatorAdded.connect(this->onOperatorAddedSlot);
	mPatch.operatorWillBeRemoved.connect(this->onOperatorRemovedSlot);

	for (auto op : mPatch.getOperators()) {
		nap::Operator* opp = op;
		onOperatorAdded(*opp);
	}
}

void PatchScene::onSelectionChanged()
{
	QList<nap::Operator*> nodes;
	for (QGraphicsItem* item : selectedItems()) {
		OperatorItem* n = qgraphicsitem_cast<OperatorItem*>(item);
		if (!n) continue;
		nodes.append(&n->getOperator());
	}

    QList<nap::Object*> objects;
    for (auto op : nodes) {
        objects << op;
    }
    AppContext::get().setSelection(objects);

	operatorSelectionChanged(nodes);
}

void PatchScene::startDragConnection(PinItem* pin)
{
	hideIncompatiblePlugs(pin->plugItem());

	mWireIsOutput = !pin->plugItem()->isInput();
	if (mWireIsOutput) {
		mPreviewWire->setSrcPin(pin);
		mPreviewWire->setDstPos(pin->attachPos());
	} else {
		mPreviewWire->setDstPin(pin);
		mPreviewWire->setSrcPos(pin->attachPos());
	}

	mPreviewWire->setVisible(true);
	mIsWiring = true;
}

void PatchScene::updateDragConnection(QPointF pt)
{
	if (mWireIsOutput) {
		mPreviewWire->setDstPos(pt);
	} else {
		mPreviewWire->setSrcPos(pt);
	}
	mPreviewWire->updatePath();
}

void PatchScene::stopDragConnection(PlugItem* plug)
{
	if (!mIsWiring) return;

	if (plug) { // Only do this when we got a pin
		PlugItem* srcPlug;
		PlugItem* dstPlug;
		if (mWireIsOutput) {
			srcPlug = mPreviewWire->srcPin()->plugItem();
			dstPlug = plug;
		} else {
			srcPlug = plug;
			dstPlug = mPreviewWire->dstPin()->plugItem();
		}

		if (srcPlug != dstPlug)
			AppContext::get().execute(
				new ConnectPlugsCmd((nap::OutputPlugBase&)srcPlug->plug(), (nap::InputPlugBase&)dstPlug->plug()));
	}
	mPreviewWire->setSrcPin(NULL);
	mPreviewWire->setDstPin(NULL);
	mPreviewWire->setVisible(false);
	mIsWiring = false;

	showAllPlugs();
}

void PatchScene::drawBackground(QPainter* painter, const QRectF& rect)
{
	qreal spacing = 50;
	int xMin = qFloor(rect.left() / spacing);
	int xMax = qCeil(rect.right() / spacing);
	int yMin = qFloor(rect.top() / spacing);
	int yMax = qCeil(rect.bottom() / spacing);

	QPen pen;
	pen.setWidth(0);
	pen.setColor(QApplication::palette().base().color().darker(110));
	painter->setPen(pen);

	for (int x = xMin; x < xMax; x++) {
		painter->drawLine(QPointF(x, yMin) * spacing, QPointF(x, yMax) * spacing);
	}
	for (int y = yMin; y < yMax; y++) {
		painter->drawLine(QPointF(xMin, y) * spacing, QPointF(xMax, y) * spacing);
	}
}

QList<OperatorItem*> PatchScene::operators() const { return filter<OperatorItem*>(mOperatorLayer->childItems()); }

QList<WireItem*> PatchScene::wires() const { return filter<WireItem*>(mWireLayer->childItems()); }

nap::Patch& PatchScene::patch() const { return mPatch; }

void PatchScene::hideIncompatiblePlugs(PlugItem* src)
{
	for (OperatorItem* node : operators())
		node->hideIncompatiblePlugs(src);
}

void PatchScene::showAllPlugs()
{
	for (OperatorItem* node : operators())
		node->showAllPlugs();
}

QList<OperatorItem*> PatchScene::selectedOperatorItems() const { return filter<OperatorItem*>(selectedItems()); }



QList<WireItem*> PatchScene::selectedWires() const { return filter<WireItem*>(selectedItems()); }

void PatchScene::removeOperatorItem(OperatorItem* item) { delete item; }

OperatorItem* PatchScene::findOperatorItem(const std::string& id) const
{
	for (OperatorItem* node : operators())
		if (node->name() == id) return node;
	return nullptr;
}

OperatorItem* PatchScene::findOperatorItem(nap::Operator& op) const { return findOperatorItem(op.getName()); }

nap::Plug& PatchScene::dragConnectionSource() const
{
	if (mPreviewWire->srcPin()) return mPreviewWire->srcPin()->plugItem()->plug();
	return mPreviewWire->dstPin()->plugItem()->plug();
}

void PatchScene::onOperatorAdded(nap::Operator& op)
{
	OperatorItem* item = new OperatorItem(mOperatorLayer, op);

	connect(item, &OperatorItem::plugConnected,
			[=](const nap::OutputPlugBase& srcPlug, const nap::InputPlugBase& dstPlug) { addWire(srcPlug, dstPlug); });

	connect(item, &OperatorItem::plugDisconnected,
			[=](const nap::OutputPlugBase& srcPlug, const nap::InputPlugBase& dstPlug) {
				assert(false); // Not implemented
			});

	item->setParentItem(mOperatorLayer);
}

void PatchScene::onOperatorRemoved(nap::Operator& op)
{
	OperatorItem* item = findOperatorItem(op);
	assert(item);
	removeOperatorItem(item);
}

void PatchScene::onOperatorChanged(nap::Operator& op) { findOperatorItem(op)->setPos(getObjectEditorPosition(op)); }


void PatchScene::addWire(const nap::OutputPlugBase& srcPlug, const nap::InputPlugBase& dstPlug)
{
	OperatorItem* srcNodeItem = findOperatorItem(*srcPlug.getParent());
	PlugItem* srcPlugItem = srcNodeItem->findPlugItem(srcPlug);

	OperatorItem* dstNodeItem = findOperatorItem(*dstPlug.getParent());
	PlugItem* dstPlugItem = dstNodeItem->findPlugItem(dstPlug);

	WireItem* item = new WireItem(srcPlugItem->pin(), dstPlugItem->pin());
	item->setParentItem(mWireLayer);
}

// void PatchScene::onWireRemoved(nap::Wire* wire) { delete findWire(wire); }
